lib/quicly.c

```diff
@@ -3677,197 +3733,137 @@ static int handle_close(quicly_conn_t *conn, int err, uint64_t frame_type, ptls_
+                         VEC_EVENT_ATTR(REASON_PHRASE, frame.reason_phrase));
+    return handle_close(conn, QUICLY_ERROR_FROM_TRANSPORT_ERROR_CODE(frame.error_code), frame.frame_type, frame.reason_phrase);
 }
 
-static int handle_application_close(quicly_conn_t *conn, uint16_t error_code, ptls_iovec_t reason_phrase)
+static int handle_application_close_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
 {
-    LOG_CONNECTION_EVENT(conn, QUICLY_EVENT_TYPE_APPLICATION_CLOSE_RECEIVE, INT_EVENT_ATTR(ERROR_CODE, error_code),
-                         VEC_EVENT_ATTR(REASON_PHRASE, reason_phrase));
-    return handle_close(conn, QUICLY_ERROR_FROM_APPLICATION_ERROR_CODE(error_code), UINT64_MAX, reason_phrase);
+    quicly_application_close_frame_t frame;
+    int ret;
+
+    if ((ret = quicly_decode_application_close_frame(&state->src, state->end, &frame)) != 0)
+        return ret;
+
+    LOG_CONNECTION_EVENT(conn, QUICLY_EVENT_TYPE_APPLICATION_CLOSE_RECEIVE, INT_EVENT_ATTR(ERROR_CODE, frame.error_code),
+                         VEC_EVENT_ATTR(REASON_PHRASE, frame.reason_phrase));
+    return handle_close(conn, QUICLY_ERROR_FROM_APPLICATION_ERROR_CODE(frame.error_code), UINT64_MAX, frame.reason_phrase);
 }
 
-static int handle_stateless_reset(quicly_conn_t *conn)
+static int handle_padding_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
 {
-    LOG_CONNECTION_EVENT(conn, QUICLY_EVENT_TYPE_STATELESS_RESET_RECEIVE);
-    return handle_close(conn, QUICLY_ERROR_RECEIVED_STATELESS_RESET, UINT64_MAX, ptls_iovec_init("", 0));
+    ++state->num_frames_non_ack_eliciting;
+    return 0;
 }
 
-static int handle_payload(quicly_conn_t *conn, size_t epoch, const uint8_t *src, size_t _len, uint64_t *offending_frame_type,
-                          int *is_ack_only)
+static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
 {
-    const uint8_t *end = src + _len;
-    uint8_t frame_type;
-    int ret = 0;
+    return 0;
+}
 
-    *is_ack_only = 1;
+static int handle_new_connection_id_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
+{
+    quicly_new_connection_id_frame_t frame;
+    return quicly_decode_new_connection_id_frame(&state->src, state->end, &frame);
+}
+
+static int handle_retire_connection_id_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
+{
+    return QUICLY_TRANSPORT_ERROR_PROTOCOL_VIOLATION;
+}
+
+static int handle_payload(quicly_conn_t *conn, size_t epoch, const uint8_t *_src, size_t _len, uint64_t *offending_frame_type,
+                          int *is_ack_only)
+{
+    static const struct {
+        int (*cb)(quicly_conn_t *, struct st_quicly_handle_payload_state_t *);
+        uint8_t permitted_epochs; /* bitmask of 1<<epoch */
+    } frame_handlers[] = {
+    /* clang-format off */
+#define FRAME(n, i, z, h, o) {handle_##n##_frame, (i ? 1 : 0) | (z ? 2 : 0) | (h ? 4 : 0) | (o ? 8 : 0)}
```

Comment:
This is clever! Can you add a comment to describe this array, and how it's populated? Giving an example might be useful, if that can be done.

Reply:
Added comments, added headers to the table.